File: benchmarks/language-comparison/programs/workloads.ts
// Bun strips types; the companion tsc check is a separate operation.
declare const process: { argv: string[] };
type Cell = Readonly<{ index: number; value: number }>;

function scalar(steps: number, value: number): number {
    for (let step = 0; step < steps; step++) value = (value * 48271) % 2147483647;
    return value;
}

function scalarBigInt(steps: number, value: bigint): bigint {
    for (let step = 0; step < steps; step++) value = (value * 48271n) % 2147483647n;
    return value;
}

function model(steps: number, seed: number, mutable: boolean): [number, number] {
    const original: readonly Cell[] = Array.from({ length: 256 }, (_, index) => ({ index, value: 0 }));
    let values = [...original];
    for (let step = 0; step < steps; step++) {
        const target = (seed + step * 17) % 256;
        if (mutable) {
            const cell = values[target];
            if (!cell) throw new Error("controlled index outside model");
            values[target] = { ...cell, value: cell.value + 1 };
        } else {
            values = values.map(cell => cell.index === target ? { ...cell, value: cell.value + 1 } : cell);
        }
    }
    return [values.reduce((sum, cell) => sum + (cell.index + 1) * cell.value, 0),
        original.reduce((sum, cell) => sum + cell.value, 0)];
}

const steps = Number(process.argv[3]);
const seed = Number(process.argv[4]);
switch (process.argv[2]) {
    case "scalar": console.log(scalar(steps, seed)); break;
    case "scalar-bigint": console.log(scalarBigInt(steps, BigInt(seed)).toString()); break;
    case "precision": console.log((9007199254740993n + BigInt(seed)).toString()); break;
    case "precision-number": console.log(9007199254740993 + seed); break;
    default: {
        const [sum, old] = model(steps, seed, process.argv[2] === "mutable");
        console.log(`${sum}\n${old}`);
    }
}

File: benchmarks/language-comparison/programs/workloads.rs
use std::env;

#[derive(Clone, Copy)]
struct Cell {
    index: i64,
    value: i64,
}

fn scalar(steps: i64, mut value: i64) -> i64 {
    for _ in 0..steps {
        value = value * 48_271 % 2_147_483_647;
    }
    value
}

fn model(steps: i64, seed: i64, mutable: bool) -> (i64, i64) {
    let original: Vec<_> = (0..256).map(|index| Cell { index, value: 0 }).collect();
    let mut values = original.clone();
    for step in 0..steps {
        let target = (seed + step * 17) % 256;
        if mutable {
            values[target as usize].value += 1;
        } else {
            // Keep every old version unmodified; allocate the next whole vector.
            values = values
                .iter()
                .map(|cell| {
                    if cell.index == target {
                        Cell {
                            value: cell.value + 1,
                            ..*cell
                        }
                    } else {
                        *cell
                    }
                })
                .collect();
        }
    }
    (
        values
            .iter()
            .map(|cell| (cell.index + 1) * cell.value)
            .sum(),
        original.iter().map(|cell| cell.value).sum(),
    )
}

fn main() {
    let args: Vec<_> = env::args().collect();
    let steps: i64 = args[2].parse().expect("controlled decimal steps");
    let seed: i64 = args[3].parse().expect("controlled decimal seed");
    match args[1].as_str() {
        "scalar" => println!("{}", scalar(steps, seed)),
        "precision" => println!("{}", 9_007_199_254_740_993_i64 + seed),
        mode => {
            let (sum, old) = model(steps, seed, mode == "mutable");
            println!("{sum}\n{old}");
        }
    }
}
